File: frontend/src/components/Resources.js
import React, { useState, useEffect } from 'react';
import { apiService, formatCurrency, formatNumber } from '../services/api';
import './Resources.css';

const Resources = () => {
  const [selectedRegion, setSelectedRegion] = useState('Abuja');
  const [allocation, setAllocation] = useState(null);
  const [inventory, setInventory] = useState({});
  const [requirements, setRequirements] = useState(null);
  const [loading, setLoading] = useState(false);
  const [error, setError] = useState(null);

  const nigerianRegions = [
    'Abuja', 'Lagos', 'Kano', 'Port Harcourt', 'Ibadan', 
    'Kaduna', 'Benin City', 'Jos', 'Maiduguri', 'Sokoto',
    'Ilorin', 'Enugu', 'Abeokuta', 'Owerri', 'Calabar'
  ];

  const resourceIcons = {
    food_packets: '🍱',
    water_bottles: '💧',
    medical_kits: '🏥',
    blankets: '🛏️',
    rescue_boats: '🚤',
    shelters: '🏕️'
  };

  useEffect(() => {
    fetchGlobalInventory();
    if (selectedRegion) {
      fetchRegionRequirements(selectedRegion);
    }
  }, [selectedRegion]);

  const fetchGlobalInventory = async () => {
    try {
      const response = await apiService.getGlobalInventory();
      setInventory(response.data.inventory || {});
    } catch (err) {
      console.error('Failed to fetch inventory:', err);
    }
  };

  const fetchRegionRequirements = async (region) => {
    try {
      const response = await apiService.getResourceRequirements(region);
      setRequirements(response.data);
    } catch (err) {
      console.error('Failed to fetch requirements:', err);
    }
  };

  const allocateResources = async (severity = 0.8) => {
    try {
      setLoading(true);
      setError(null);
      
      const response = await apiService.allocateResources(
        selectedRegion,
        null, // Let the system estimate population
        severity
      );
      
      setAllocation(response.data.allocation);
      await fetchGlobalInventory(); // Refresh inventory
      
    } catch (err) {
      console.error('Failed to allocate resources:', err);
      setError('Failed to allocate resources');
    } finally {
      setLoading(false);
    }
  };

  const handleRegionChange = (region) => {
    setSelectedRegion(region);
    setAllocation(null); // Clear previous allocation
  };

  return (
    <div className="resources-page">
      <div className="resources-header">
        <h1>Resource Management</h1>
        <p>Allocate and track emergency resources across regions</p>
      </div>

      <div className="grid grid-cols-3">
        {/* Region Selector & Actions */}
        <div className="card">
          <div className="card-header">
            <h2 className="card-title">Select Region</h2>
          </div>
          
          <div className="region-selector">
            <select
              value={selectedRegion}
              onChange={(e) => handleRegionChange(e.target.value)}
              className="form-control region-select"
            >
              {nigerianRegions.map(region => (
                <option key={region} value={region}>{region}</option>
              ))}
            </select>
          </div>

          <div className="allocation-actions">
            <button
              className="btn btn-primary allocation-btn"
              onClick={() => allocateResources(0.5)}
              disabled={loading}
            >
              {loading ? 'Allocating...' : 'Standard Allocation'}
            </button>
            
            <button
              className="btn btn-danger allocation-btn"
              onClick={() => allocateResources(0.9)}
              disabled={loading}
            >
              {loading ? 'Allocating...' : 'Emergency Allocation'}
            </button>
          </div>

          {error && (
            <div className="error">
              <p>{error}</p>
            </div>
          )}
        </div>

        {/* Resource Requirements */}
        <div className="card">
          <div className="card-header">
            <h2 className="card-title">Requirements - {selectedRegion}</h2>
          </div>
          
          {requirements ? (
            <div className="requirements-content">
              <div className="requirements-summary">
                <div className="summary-item">
                  <label>Population</label>
                  <value>{formatNumber(requirements.population)}</value>
                </div>
                <div className="summary-item">
                  <label>Estimated Cost</label>
                  <value>{formatCurrency(requirements.estimated_cost)}</value>
                </div>
                <div className="summary-item">
                  <label>Priority</label>
                  <value className={`priority-${requirements.priority}`}>
                    {requirements.priority.toUpperCase()}
                  </value>
                </div>
              </div>

              <div className="requirements-list">
                {Object.entries(requirements.requirements || {}).map(([resource, quantity]) => (
                  <div key={resource} className="requirement-item">
                    <span className="resource-icon">
                      {resourceIcons[resource] || '📦'}
                    </span>
                    <span className="resource-name">
                      {resource.replace('_', ' ').replace(/\b\w/g, l => l.toUpperCase())}
                    </span>
                    <span className="resource-quantity">
                      {formatNumber(quantity)}
                    </span>
                  </div>
                ))}
              </div>
            </div>
          ) : (
            <div className="loading">Loading requirements...</div>
          )}
        </div>

        {/* Current Allocation */}
        <div className="card">
          <div className="card-header">
            <h2 className="card-title">Current Allocation</h2>
          </div>
          
          {allocation ? (
            <div className="allocation-content">
              <div className="allocation-summary">
                <div className="summary-metric">
                  <h3>{formatNumber(allocation.affected_population)}</h3>
                  <p>Affected Population</p>
                </div>
                <div className="summary-metric">
                  <h3>{formatCurrency(allocation.total_cost_estimate)}</h3>
                  <p>Total Cost</p>
                </div>
                <div className="summary-metric">
                  <h3 className={`priority-${allocation.priority}`}>
                    {allocation.priority.toUpperCase()}
                  </h3>
                  <p>Priority Level</p>
                </div>
              </div>

              <div className="allocated-resources">
                {Object.entries(allocation.resources || {}).map(([resource, quantity]) => (
                  <div key={resource} className="allocated-item">
                    <span className="resource-icon">
                      {resourceIcons[resource] || '📦'}
                    </span>
                    <div className="resource-details">
                      <span className="resource-name">
                        {resource.replace('_', ' ').replace(/\b\w/g, l => l.toUpperCase())}
                      </span>
                      <span className="resource-quantity">
                        {formatNumber(quantity)} units
                      </span>
                    </div>
                  </div>
                ))}
              </div>
            </div>
          ) : (
            <div className="no-allocation">
              <p>Select allocation type above to see resource distribution</p>
            </div>
          )}
        </div>
      </div>

      {/* Global Inventory */}
      <div className="card">
        <div className="card-header">
          <h2 className="card-title">Global Resource Inventory</h2>
          <button 
            className="btn btn-secondary"
            onClick={fetchGlobalInventory}
          >
            Refresh
          </button>
        </div>
        
        <div className="inventory-grid">
          {Object.keys(inventory).length > 0 ? (
            Object.entries(inventory).map(([region, resources]) => (
              <div key={region} className="inventory-region">
                <h3 className="region-name">{region}</h3>
                <div className="region-resources">
                  {Object.entries(resources).map(([resourceType, data]) => (
                    <div key={resourceType} className="inventory-item">
                      <span className="resource-icon">
                        {resourceIcons[resourceType] || '📦'}
                      </span>
                      <div className="inventory-details">
                        <span className="resource-name">
                          {resourceType.replace('_', ' ').replace(/\b\w/g, l => l.toUpperCase())}
                        </span>
                        <span className="resource-stats">
                          Allocated: {formatNumber(data.total_allocated || 0)}
                        </span>
                      </div>
                    </div>
                  ))}
                </div>
              </div>
            ))
          ) : (
            <div className="no-inventory">
              <p>No inventory data available. Allocate resources to regions to see inventory.</p>
            </div>
          )}
        </div>
      </div>
    </div>
  );
};

export default Resources;

File: backend/routes/resources.py
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel
from typing import Dict, Any, List
from services.resource_allocator import resource_allocator

router = APIRouter()

class ResourceAllocationRequest(BaseModel):
    region: str
    affected_population: int = None
    severity: float = 0.5

class ResourceUpdateRequest(BaseModel):
    region: str
    resource_type: str
    quantity: int
    action: str = "allocate"  # allocate, distribute, request

@router.post("/allocate")
def allocate_resources(request: ResourceAllocationRequest) -> Dict[str, Any]:
    """
    Allocate resources for a region based on flood prediction
    
    Args:
        request: Resource allocation request
        
    Returns:
        Resource allocation details
    """
    try:
        allocation = resource_allocator.allocate_resources(
            region=request.region,
            affected_population=request.affected_population,
            severity=request.severity
        )
        
        if "error" in allocation:
            raise HTTPException(status_code=500, detail=allocation["error"])
        
        return {
            "status": "success",
            "allocation": allocation
        }
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Resource allocation error: {str(e)}")

@router.get("/status/{region}")
def get_resource_status(region: str) -> Dict[str, Any]:
    """
    Get current resource status for a region
    
    Args:
        region: Region name
        
    Returns:
        Current resource allocation and availability
    """
    try:
        status = resource_allocator.get_resource_status(region)
        
        return {
            "status": "success",
            "region": region,
            "resources": status
        }
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to get resource status: {str(e)}")

@router.post("/update")
def update_resource_inventory(update: ResourceUpdateRequest) -> Dict[str, Any]:
    """
    Update resource inventory (allocate, distribute, or request more)
    
    Args:
        update: Resource update request
        
    Returns:
        Update confirmation
    """
    try:
        from database import supabase
        
        # Record the resource update
        update_record = {
            "region": update.region,
            "resource_type": update.resource_type,
            "quantity": update.quantity,
            "action": update.action
        }
        
        result = supabase.table("resources").insert(update_record).execute()
        
        return {
            "status": "success",
            "update": update_record,
            "database_id": result.data[0]["id"] if result.data else None
        }
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to update resources: {str(e)}")

@router.get("/inventory")
def get_global_inventory() -> Dict[str, Any]:
    """
    Get global resource inventory across all regions
    
    Returns:
        Global resource status
    """
    try:
        from database import supabase
        
        # Get all resource records
        result = supabase.table("resources").select("*").execute()
        
        # Aggregate by region and resource type
        inventory = {}
        for record in result.data:
            region = record["region"]
            resource_type = record["resource_type"]
            
            if region not in inventory:
                inventory[region] = {}
            
            if resource_type not in inventory[region]:
                inventory[region][resource_type] = {
                    "total_allocated": 0,
                    "total_distributed": 0,
                    "available": 0
                }
            
            # This is a simplified aggregation - in practice, you'd want more sophisticated tracking
            inventory[region][resource_type]["total_allocated"] += record["quantity"]
        
        return {
            "status": "success",
            "inventory": inventory,
            "total_regions": len(inventory)
        }
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to get inventory: {str(e)}")

@router.get("/requirements/{region}")
def get_resource_requirements(region: str, population: int = None) -> Dict[str, Any]:
    """
    Get resource requirements for a region based on population
    
    Args:
        region: Region name
        population: Affected population (optional)
        
    Returns:
        Resource requirements breakdown
    """
    try:
        if not population:
            # Estimate population based on region
            regional_populations = {
                "Abuja": 500000,
                "Lagos": 2000000,
                "Kano": 800000,
                "Port Harcourt": 600000,
                "Ibadan": 700000,
                "Kaduna": 400000
            }
            population = regional_populations.get(region, 100000)
        
        # Calculate requirements using the resource allocator logic
        allocation = resource_allocator.allocate_resources(region, population, 0.8)  # High severity
        
        if "error" in allocation:
            raise HTTPException(status_code=500, detail=allocation["error"])
        
        return {
            "status": "success",
            "region": region,
            "population": population,
            "requirements": allocation["resources"],
            "estimated_cost": allocation["total_cost_estimate"],
            "priority": allocation["priority"]
        }
        
    except HTTPException:
        raise
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to get requirements: {str(e)}")

